posts/inside-rust/2021-05-15-What-the-error-handling-project-group-is-working-towards.md

```diff
@@ -0,0 +1,235 @@
+
+> Disclaimer: *This post is equal parts plan and aspiration. There are technical challenges here to sort out so the final outcome may look rather different from our initial vision, so please don't assume any of this is final.*
+
+## Error Handling Today
+
+The first challenge we'd like to solve is that it's easy to lose context accidentally when reporting errors. There are a couple of places this can happen, either when printing an error and forgetting to print sources, when returning an error from main, or when converting a recoverable error into a non recoverable error.
+
+Consider this example:
+
+```rust
+use std::fmt;
+
+// We have a program that loads a config and expects that
+// loading the config will always succeed.
+fn main() {
+    let _config = load_config()
+        .expect("config is always valid and exists");
+}
+
+// We have a dummy implementation of load_config which
+// always errors, since we're just focusing on diagnostics
+// here.
+fn load_config() -> Result<(), Error> {
+    Err(Error(SourceError))
+}
+
+// And we have an error type that just prints "invalid
+// config" and has a source error which just prints "config
+// file does not exist"
+#[derive(Debug)]
+struct Error(SourceError);
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("invalid config")
+    }
+}
+
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        Some(&self.0)
+    }
+}
+
+#[derive(Debug)]
+struct SourceError;
+
+impl fmt::Display for SourceError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("config file does not exist")
+    }
+}
+
+impl std::error::Error for SourceError {}
+```
+
+When we run this we would like to see output somewhat like this:
+
+```
+$ cargo run
+thread 'main' panicked at 'config is always valid and exists', src/main.rs:4:33
+
+Error:
+    0: invalid config
+    1: config file does not exist
+
+note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+```
+
+In this error message we can see that we exited because of a panic. We can see what invariant we violated that was supposed to prevent this panic from happening. We can see the location where the panic was produced. And we can see the error message of each error in the chain of errors accessible via `source`.
+
+That's what we would like, at least in the version of rust that the error handling project group wants to see, but what we actually get is this...
```

Comment:
capital Rust or lowercase rust? Is there a consensus for using either naming?

Reply:
should be capital everywhere, :+1: